web-runner.go

```diff
@@ -0,0 +1,237 @@
+package web
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"github.com/launchrctl/launchr/pkg/cli"
+	"github.com/launchrctl/launchr/pkg/log"
+
+	"github.com/launchrctl/web/server"
+)
+
+const (
+	backgroundCmdEnvVar = "BACKGROUND_CMD"
+)
+
+func isBackGroundEnv() bool {
+	return len(os.Getenv(backgroundCmdEnvVar)) == 1
+}
+
+func runWeb(ctx context.Context, p *Plugin, webOpts *webFlags) error {
+	var err error
+
+	port := webOpts.Port
+	if !server.IsPortFree(port) {
+		log.Info("The port %d you are trying to use for the web server is not available.", port)
+		port, err = server.GetAvailablePort(port)
+		if err != nil {
+			return err
+		}
+	}
+
+	serverOpts := &server.RunOptions{
+		Addr:          fmt.Sprintf(":%d", port), // @todo use proper addr
+		APIPrefix:     APIPrefix,
+		SwaggerJSON:   webOpts.UseSwaggerUI,
+		ProxyClient:   webOpts.ProxyClient,
+		RunInfoFolder: webOpts.RunInfoFolder,
+		// @todo use embed fs for client or provide path ?
+	}
+
+	// @todo to consider renaming and removing access to plugin and overall global assets.
+	prepareRunOption(p, serverOpts)
+
+	return server.Run(ctx, p.app, serverOpts)
+}
+
+func runBackgroundWeb(cmd *cobra.Command, p *Plugin, flags *webFlags, pidFile string) error {
+	if isBackGroundEnv() {
+		err := redirectOutputs(flags.RunInfoFolder)
+		if err != nil {
+			return err
+		}
+
+		return runWeb(cmd.Context(), p, flags)
+	}
+
+	pid, err := runBackgroundCmd(cmd, pidFile)
+	if err != nil {
+		return err
+	}
+
+	// Wait until background server is up.
+	// Check if run info created and server is reachable.
+	// Print server URL in CLI.
+	// Kill process in case of timeout
+	timeout := time.After(10 * time.Second)
+	ticker := time.NewTicker(1 * time.Second)
+
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-timeout:
+			// Kill existing process
+			_ = killProcess(pid)
+
+			// Cleanup temp folder
+			err = os.RemoveAll(flags.RunInfoFolder)
+			if err != nil {
+				log.Debug(err.Error())
+			}
+
+			return errors.New("couldn't start background process")
+		case <-ticker.C:
+			runInfo, _ := server.GetRunInfo(flags.RunInfoFolder)
+			if runInfo == nil {
+				continue
+			}
+
+			cli.Println("Web running in background with PID:%d", pid)
+			if runInfo.BaseURL != "" {
```

Comment:
Can it really be empty? If it's empty, we must panic, the previous process failed.